packages/@aws-cdk/pipelines/lib/main/engine.ts

```diff
@@ -0,0 +1,30 @@
+import { Construct } from 'constructs';
+import { Blueprint } from '../blueprint';
+
+/**
+ * A CDK Pipelines deployment engine
+ *
+ * Implement a CDK Deployment engine by implementing this interface.
+ */
+export interface IDeploymentEngine {
+  /**
+   * Called when the pipeline is configured and the deployment should to be created
+   */
+  buildDeployment(options: BuildDeploymentOptions): void;
+}
+
+
+/**
+ * Options for the `IDeploymentEngine.buildDeployment()` calls.
+ */
+export interface BuildDeploymentOptions {
+  /**
+   * A scope where new constructs can be added
+   */
+  readonly scope: Construct;
```

Comment:
If engines were constructs, this wouldn't be needed right? Would that make more sense?

Reply:
Engines were constructs before but the construction order doesn't come out right. You can't both construct them under `this` *and* pass them to `super()`, which is an issue if you want to be able to say `CodePipelinePipeline` is-a `Pipeline`, which we kinda do.

Reply:
As long as we can offer a the same API (and reuse the graph implementation) and support multiple engines, I am comfortable hiding the engine implementation.

I think `Pipeline` (which I think we should merge with `Blueprint`) should be construct so that it's possible to simply do:

```ts
class CodePipelinePipeline extends Pipeline {
  construct(scope: Construct, id: string, props: CodePipelinePipelineProps) {
    super(scope, id);
  }

  public buildDeployment() {
    // ...
  }
}
```